src/engine/Settings.cs

```diff
@@ -395,7 +455,106 @@ private void CopySettings(Settings settings)
             if (!property.CanWrite)
                 continue;
 
-            property.SetValue(this, property.GetValue(settings));
+            // Since the properties we want to copy are SettingValue generics we use the IAssignableSetting
+            // interface and AssignFrom method to convert the property to the correct concrete class.
+            var setting = property.GetValue(this) as IAssignableSetting;
+
+            setting?.AssignFrom(property.GetValue(settings));
+        }
+    }
+
+    /// <summary>
+    ///   Wrapper class for settings options containing the value and a delegate that provides a callback for
+    ///   when the value is changed.
+    /// </summary>
+    public class SettingValue<TValueType> : IAssignableSetting
+    {
+        private TValueType value;
+
+        public SettingValue(TValueType value)
+        {
+            this.value = value;
+        }
+
+        public event SettingValueChangedDelegate<TValueType> OnChanged;
+
+        public TValueType Value
+        {
+            get => value;
+            set
+            {
+                if (!Equals(this.value, value))
+                {
+                    this.value = value;
+
+                    OnChanged?.Invoke(value);
+                }
+            }
+        }
+
+        public static implicit operator TValueType(SettingValue<TValueType> value)
+        {
+            return value.value;
+        }
+
+        public static bool operator ==(SettingValue<TValueType> lhs, SettingValue<TValueType> rhs)
+        {
+            return Equals(lhs, rhs);
+        }
+
+        public static bool operator !=(SettingValue<TValueType> lhs, SettingValue<TValueType> rhs)
+        {
+            return !(lhs == rhs);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null)
+            {
+                return false;
+            }
+
+            if (GetType() != obj.GetType())
+            {
+                return false;
+            }
+
+            return Equals((SettingValue<TValueType>)obj);
+        }
+
+        public bool Equals(SettingValue<TValueType> obj)
+        {
+            if (!value.Equals(obj.value))
+                return false;
+
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
+            return 17 ^ value.GetHashCode();
+        }
+
+        /// <summary>
+        ///   Casts a parameter object into a SettingValue generic of a matching type (if possible) and then
+        ///   copies the value from it.
+        /// </summary>
+        public void AssignFrom(object obj)
+        {
+            // Convert the object to the correct concrete type if possible.
+            SettingValue<TValueType> settingObject = obj as SettingValue<TValueType>;
```

Comment:
`var settingObject`
(github didn't let me submit this comment previously as it complained the diff was outdated)